src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -42,25 +49,68 @@ public Product getProduct(Long id) {
             return products.stream()
                     .filter(product -> id.equals(product.getId()))
                     .findAny()
-                    .orElseThrow(ProductNotFoundException::new);
+                    .orElseThrow(()->new ProductNotFoundException(id));
         } finally {
             readLock.unlock();
         }
     }
 
     @Override
-    public List<Product> findProducts() {
+    public List<Product> findProducts(String query, SortField sortField, SortType sortType) {
         readLock.lock();
         try {
             return products.stream()
+                    .filter(product ->
+                                    query == null ||
+                                    query.equals("") ||
+                                    search(query, product))
                     .filter(product -> product.getStock() > 0)
                     .filter(product -> product.getPrice() != null)
+                    .sorted((p1, p2) -> relevanceSort(query, p1, p2))
+                    .sorted(getSortComparator(sortField, sortType))
                     .collect(Collectors.toList());
         } finally {
             readLock.unlock();
         }
     }
 
+    private boolean search(String query, Product product) {
+        return Arrays.stream(query.split("\\s+"))
+                .anyMatch(word -> product.getDescription().toLowerCase()
+                        .contains(word.toLowerCase()));
+    }
+
+
+    private int relevanceSort(String query, Product p1, Product p2) {
+        if (query == null || query.equals("")) return 0;
+        return Long.compare(getNumberOfCollisions(query, p2), getNumberOfCollisions(query, p1));
+    }
+
+    private long getNumberOfCollisions(String query, Product product) {
+        return Arrays.stream(query.split("\\s+"))
+                .filter(word -> product.getDescription().toLowerCase()
+                        .contains(word.toLowerCase()))
+                .count();
+    }
```

Comment:
I don't like here that for every comparison for each compared object you split query, filter and count. This can be done way more effective